src/lem/mod.rs

```diff
@@ -172,20 +287,22 @@ impl Func {
     pub fn check(&self) -> Result<()> {
         // Check if variable has already been defined. Panics
         // if it is repeated (means `deconflict` is broken)
+        use std::collections::{HashMap, HashSet};
         #[inline(always)]
-        fn is_unique(var: &Var, map: &mut HashSet<Var>) {
-            if !map.insert(var.clone()) {
+        fn is_unique(var: &Var, map: &mut HashMap<Var, bool>) {
+            if map.insert(var.clone(), false).is_some() {
                 panic!("Variable {var} already defined. `deconflict` implementation broken.");
             }
         }
-        // Check if variable is bound
+        // Check if variable is bound and sets it as "used"
         #[inline(always)]
-        fn is_bound(var: &Var, map: &HashSet<Var>) -> Result<()> {
-            map.get(var)
-                .context(format!("Variable {var} is unbound."))?;
+        fn is_bound(var: &Var, map: &mut HashMap<Var, bool>) -> Result<()> {
+            if map.insert(var.clone(), true).is_none() {
+                bail!("Variable {var} is unbound.");
+            }
```

Comment:
Doesn't this have the side effect of marking `var` as bound, even if it wasn't before? Also, separately, what's the benefit of using a `HashMap` to `bool` rather than `HashSet`?

Reply:
Okay, I'm guessing the idea (haven't seen this yet in the code) is that first you insert the var with value `false`, and then you 'upgrade' it to `true`. If it's unbound, then that it gets set to true doesn't matter because the error is unrecoverable. If that's right, I guess I get it, but it's a little funky and seems error-prone.

Reply:
Now I'm looking above, and I see the case in which `false` is inserted. It's not at all clear to me (yet) how this map is supposed to be used.

Reply:
I think maybe I'm misreading these names. `is_bound` and `is_unique` read like predicates in Rust's naming convention. But I think these functions are trying to *declare* that the var is bound/unique.

Reply:
Okay, let me check my understanding:

If, on the first occurrence of a variable, you call `is_unique` on it, then there should be no panic, and `false` will be inserted.

Then, if subsequently, you call `is_bound` on the variable, since it *has* been inserted (although with the value `false`), the `bail!` will not be triggered. And the value will be changed from `false` to `true`.

This seems to mean that a variable that was previously checked for uniqueness is now seen to be bound.

Effectively, `is_unique` means 'mark this var as being bound to an initial value', and `is_bound` means, 'mark this var has having been looked-up'.

Is that right?

Is there a significance to the panic vs bail (error) behavior?

Reply:
Oh the names are confusing indeed. It used to be that `is_bound` would just check that it was in the set, and `is_unique` would both look for it, failing if it was already in the set and mark it as bound otherwise. It used to be a `HashSet` but then I decided to add a check that all variables must be used unless it was prepended with "_". The way I do this is whenever I create a new variable in the map (by the use of is_unique) it starts as false, and any call to `is_bound` will both check and flip the flag so that it is used. This was to catch bugs in the eval function, and it did indeed catch a bunch of little bugs.